### PR TITLE
make request cancel external workflow return future

### PR DESCRIPTION
**Pull request by wxing1292**
will wait until service side change complete and use canary for testing

solves #326

### CODE REVIEW

File: internal/internal_event_handlers.go [hunk 1]
```diff
@@ -227,27 +240,25 @@ func (wc *workflowEnvironmentImpl) Complete(result []byte, err error) {
...
 		if childWorkflow.handled {
-			return nil
-		}
-		if childWorkflow.workflowExecution != nil && (decision.isDone() || !childWorkflow.waitForCancellation) {
+			// already have the result of child workflow execution, so the cancellation request is a success (bypass)
+			callback(nil, nil)
+		} else if childWorkflow.workflowExecution != nil && (decision.isDone() || !childWorkflow.waitForCancellation) {
+			// child workflow execution is cancelled, so the cancellation request is a success (bypass)
 			childWorkflow.handle(nil, ErrCanceled)
+			callback(nil, nil)
```
Comment (yiminc) [inline internal/internal_event_handlers.go]: same as above, do not call callback until you see the event from server. you might need to save the callback function to the childWorkflow instance.
also, there is an else case where you don't call callback() and the future will block forever.

File: internal/internal_event_handlers.go [hunk 2]
```diff
@@ -227,27 +240,25 @@ func (wc *workflowEnvironmentImpl) Complete(result []byte, err error) {
...
 	isChild, decision := wc.decisionsHelper.requestCancelExternalWorkflowExecution(domainName, workflowID, runID)
 	if isChild {
 		// this is for child workflow
 		childWorkflow := decision.getData().(*scheduledChildWorkflow)
 		if childWorkflow.handled {
-			return nil
-		}
-		if childWorkflow.workflowExecution != nil && (decision.isDone() || !childWorkflow.waitForCancellation) {
+			// already have the result of child workflow execution, so the cancellation request is a success (bypass)
+			callback(nil, nil)
```
[the remark this answers is not in the document]
Comment (yiminc) [reply internal/internal_event_handlers.go]: i think you should not call the callback() here. you should wait until you receive events from server and based on the event type to determine if it the request is sent or failed.

File: internal/internal_workflow_testsuite.go [hunk 3]
```diff
@@ -1206,9 +1207,10 @@ func (env *testWorkflowEnvironmentImpl) RequestCancelWorkflow(domainName, workfl
 		// current workflow is a parent workflow, and we are canceling a child workflow
 		childHandle.handled = true
 		childEnv := childHandle.env
-		childEnv.cancelWorkflow()
+		childEnv.cancelWorkflow(callback)
 	}
-	return nil
+
+	// TODO add test for cancel external workflow
```
Comment (yiminc) [inline internal/internal_workflow_testsuite.go]: could you create an issue and put the issue number here so we don't forget

File: internal/internal_event_handlers.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (wxing1292) [reply internal/internal_event_handlers.go]: just FYI, this PR has not change any behavior yet, i am still fixing bugs on both the client and server side for cancellation.

**Pull request closed without merging**